src/modules/strings.wren

```diff
@@ -12,4 +12,57 @@ class Strings {
             return String.fromByte(x)
         }.join("")
     }
+
+    /* String matching with wildcards '*' and '?'
+     * Algorithm from https://www.geeksforgeeks.org/wildcard-pattern-matching/
+     */
+    static strmatch(s, pattern) {
```

Comment:
Is this a common string matching utility function?  Is there any justification for the naming `strmatch` here, or just the first thing you thought of?  I'm too familiar with regex to know much of these things.

Reply:
In my research, it seems to be a common C function name for string matching using glob patterns.

I did think about this, but couldn't quickly think of a better name:
* `match` or `matches` should be reserved for Regex matching
* `globMatch` is kind of ugly

I'm definitely open to suggestions.

My first real language was Tcl. It has full regex support, but also a handy [`string match`](http://www.tcl-lang.org/man/tcl8.6/TclCmd/string.htm) command, using glob patterns, for when you don't need the full power of a regex.

My thinking with this PR is that it will eventually turn into
```
class Directory {
  //...
  static glob(dir, pattern) {
    list(pattern).where {|file| Strings.strmatch(file, pattern)}.toList
  }
}
```

Reply:
Not sure I follow did you mean `list(dir)`?  Where is the list of potential matches?

Reply:
https://github.com/joshgoebel/wren-console/blob/main/src/module/io.wren -- Directory.list(path)

Reply:
Ah yes, but `list` knows nothing about your magic pattern... it takes a path, not a pattern.

Reply:
To me a basic glob implementation has to at least deal with`*` and `**` which means it has to know a lot about dir structure, it can't just be a "scan all and filter"...  At least those are the two things I use with glob most commonly.

Reply:
I know. After we resolve this PR, the next one will introduce the Directory.glob(path, pattern) method.

Reply:
I'm not (yet) sure we should have glob in `Directory`... now we're getting into Console vs CLI and whether we want to implement things we can NEVER upstream... CLI has been very minimal/conservative in the past... and Essentials is compiled into wren-console (but may never be in CLI).  So making the core library functions suddenly dependent on Essentials to function means we kind of permanently break their eligibility for CLI porting.

So, for now, why not add a `Glob` class to the `essentials` module... and then this PR can simply be expanded to include that work as well.

Reply:
OK, I'm rethinking the glob stuff. I'll step back and look at making Directory a more "first class" class with a C implementation.

However, I still think Strings.strmatch (pending a better name) is a useful function until someone (could be me) implements regular expressions.

Reply:
>  I'll step back and look at making Directory a more "first class" class with a C implementation.

I'm not sure I know what you mean?  Much of the stuff in Directory/File is already C via `libuv`, no?

>  until someone (could be me) implements regular expressions.

Like natively?  I'd imagine the best way to do regex would be via a cross-platform regex C library with a nice wrapper...  (and mind want to look into the Wren repo for any history of issues related to regex support) although we're still a little crippled without native regex support in the language itself, but perhaps not terribly so.

Reply:
If I get serious about it, I'll look at wrapping either PCRE or the ["un-Tcl'ed" Tcl regex engine](https://github.com/garyhouston/hsrex). The Tcl one doesn't have _all_ the fancy features, but it looks easier to wrap and it's fast.

Reply:
My 2 cents would be that this can be on its own class.

`Glob.match`

and later if regex is available

`Regex.match`

👍

Reply:
Regex sounds right as a class, absolutely. Agree.

Glob I think is more complicated, here are a few things:

- rough glob a string (strmatch is just a lower-level foundation IMHO)
- stronger (more feature-ful) glob of a string... like `*/**/file.{js|rb}`
- Directory.glob (find all files matching a more complex pattern)

So does the latter belong in `Directory.glob` like we have `Directory.list` now?  And `Glob.match` bothers me a little because I see `strmatch` as a very early start at this stuff - a useful tool, but not the final implementation.  I was slowly leaning in towards `Strings.smatch` (simple match)... since it is a string function and I had originally imagined that string utility functions would go in the Strings domain.

So...

- `Strings.smatch`
- `Glob.match`
- other?

I don't like `strmatch` because str meant string, right? And it's already a static function on String...

@glennj Thoughts?

Reply:
I think Directory.glob can exist alongside Glob since that would compose a list of files and directories with a Glob object.

- var items = Directory.ls(path) 
- var filtered = Glob.list(items)
- filtered.simple(pattern) or
- filtered.full(pattern)

Directory.glob() would compose those

For the string can be similar. 

- var filtered = Glob.string(string)

- filtered.simple(pattern)

Reply:
My intention with this method is to introduce glob patterns into the language, starting with  a String(s) method. The next step would be to add a `glob` method to `Directory` for finding a subset of files. And Regex is a separate class (it's really a separate matter entirely).

It seems awkward to me to have a Glob class that would match against both strings and files. That class doesn't feel like it has a clear purpose. When you think of Regex, you don't think of it in terms of matching files. 

To go back to an earlier point: I don't think of `**` as part of a "simple" glob pattern. Perhaps I've been conditioned by bash where an option has to be set to enable it ([3.5.8.1 Pattern Matching](https://www.gnu.org/software/bash/manual/bash.html#Pattern-Matching)). But using glob patterns to match a string, where `*` means just "zero or more characters" doesn't require `**`: it's the `Directory` glob that needs `**` to descend recursively down directories.

Reply:
>  But using glob patterns to match a string, where * means just "zero or more characters" doesn't require **: it's the Directory glob that needs ** to descend recursively down directories.

True, but you could also imagine such a string function that said "would this path match this glob"... but for performance reasons it's easy to deal with `**` as a special case in the Directory engine itself.

So my vote this is `Strings.smatch` then (for simple match), reserving `match` for later use.